### ISSUE TITLE
Can't seem to access map or view instances after updating component with data

### ISSUE THREAD

**Issue by Chonguis** [Esri/esri-loader#91]
code link https://gist.github.com/Chonguis/348ac0914c86bf5da8fb24da7d9383a2

Hey everyone. I've been playing with react for about a month now but have worked with the arcgis js api for almost a year. I'm having a problem trying to update the map or view instances once I pass data to the esrMapiComponent. Let's say I'm passing a data object with coordinates. The initial state of the esriMapComponent has a default view on brooklyn. The user then submits a search that makes an ajax request to an API and this API responds with coordinates. Once I have this data object, I return the esriMapComponent passing in the data as props. In this, I am successful and doing a simple log on the render method will indicate that data indeed was passed to the esriMapComponent. The problem is that since the esri modules are loaded in the componentDidMount lifecycle hook, nothing seems to run again and I can't access the view or map instances created at initial state. I then want to go to view.goTo(coordinates) but I'm having trouble because if I put a console log inside of this lifecycle hook, it won't run again when the component is updated by passing the data from parent component (Map route).

My questions are: Is there another place where I could have access to the view, map, scenelayer instances that are not enclosed in this lifecycle hook? Is this problem specific to this library (react starter kit with esri-loader). Is there a better way to achieve what I want to do using the loader maybe in another way?

Hope example is clear and that this can be fixed (crossing fingers). Thank you in advance!

**Comment by andygup** [comment]
@odoe :point_up: any ideas?

**Comment by davetimmins** [comment]
@Chonguis store your view instance in the component state in `componentDidMount`, then in `componentWillReceiveProps` you can get the instance and do the updates you want.

**Comment by tomwayson** [comment]
Thanks @davetimmins!

I was thinking something along the same lines, but rather than store the `view` instance in the component's `state`, I was thinking of just setting it as a property (not to be confused with a `prop`) on the component. So [this line](https://gist.github.com/Chonguis/348ac0914c86bf5da8fb24da7d9383a2#file-esrimapcomponent-js-L27) would become `this.view = new SceneView({` and later in `componentWillReceiveProps()` you can check for `this.view` or `this.map` and make update accordingly.

I made [a branch of esri-loader-react-starter-kit that shows how to do something similar to update a 3.x) map's zoom level](https://github.com/tomwayson/esri-loader-react-starter-kit/commit/61e7cd8f811b5fe04f64660bf535950b676234a6
).

Keep in mind that I don't use React day to day, so I'm not sure if that follows React best practices, but it works.

**Comment by davetimmins** [comment]
Yea that's probably better as you don't need to trigger a re-render. I tend to have any map/view logic split out as pure functions that are somewhere outside of components and then pass the view into those so it'd more obvious where the changes are happening i.e. nothing to do with React

**Comment by tomwayson** [comment]
In other frameworks like AngularJS and Ember, I've used singleton services that get injected into the components as way of keeping a persistent connection to the map alive and managing it's state in one place. 

I'm not totally sure what the equivalent would be for React. I haven't tried this yet but I really like the way https://github.com/Esri/react-redux-js4 wraps the JSAPI in redux middleware.

**Comment by Chonguis** [comment]
@tomwayson @davetimmins thank you guys, this worked great! I stored the view instance as a property of the component. Then in componentDidUpdate() I updated the app. This worked for me better than updating it from the componentWillRecieveProps lifecycle hook because it was obligating me to re submit the information for the view to update, something that did not happen with in componentDidUpdate(). Anyways, thank you both for the input, greatly appreciate it!
